File: huongB_devirt19/GhidraFindSlot16.java
// Ghidra headless GhidraScript (Java — không cần PyGhidra).
// Khoanh vùng slot16 producer: unhex decoder @0x891f4 có nhiều caller;
// hàm nào build chuỗi 32-hex feed vào unhex = producer.
// Chạy qua run_ghidra_slot16.sh.
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.symbol.Reference;
import java.io.File;
import java.io.FileWriter;
import java.util.*;

public class GhidraFindSlot16 extends GhidraScript {
    static final long UNHEX = 0x891f4L;
    static final long SM3_DRV = 0x9fdacL;                 // consumer slot16 (SM3 driver)
    static final long[] CAND = {0x10e224L,0x10b010L,0x118ffcL};  // file offsets: const-key cipher + key-derive(SHA1?) + key-init
    static final String OUTDIR =
        "/Users/lemanh/Documents/Tiktok-Android/huongB_devirt19/_ghidra_out";

    Address A(long off){ return currentProgram.getImageBase().add(off); }

    Function fnAt(long off){
        return currentProgram.getFunctionManager().getFunctionContaining(A(off));
    }

    @Override
    public void run() throws Exception {
        new File(OUTDIR).mkdirs();
        FunctionManager fm = currentProgram.getFunctionManager();
        DecompInterface dec = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        dec.setOptions(opts);
        dec.toggleCCode(true);
        dec.toggleSyntaxTree(true);
        dec.setSimplificationStyle("decompile");
        boolean opened = dec.openProgram(currentProgram);
        println("=== decompiler openProgram = " + opened
            + (opened? "" : "  ERR: " + dec.getLastMessage()));

        println("=== image base = " + currentProgram.getImageBase());

        // callers của UNHEX (mọi caller, không chỉ 5 tĩnh — để chắc)
        println("=== callers of UNHEX 0x" + Long.toHexString(UNHEX) + " ===");
        Set<Long> dynCallers = new LinkedHashSet<Long>();
        for (Reference ref : getReferencesTo(A(UNHEX))) {
            if (ref.getReferenceType().isCall()) {
                Function fc = fm.getFunctionContaining(ref.getFromAddress());
                long ep = (fc!=null)? fc.getEntryPoint().subtract(currentProgram.getImageBase()) : -1;
                if (fc!=null) dynCallers.add(ep);
                println("  call@" + ref.getFromAddress()
                    + "  in fn 0x" + (fc!=null? Long.toHexString(ep):"?"));
            }
        }

        // gộp 5 ứng viên tĩnh + caller Ghidra tìm được
        LinkedHashSet<Long> cands = new LinkedHashSet<Long>();
        for (long c : CAND) cands.add(c);
        cands.addAll(dynCallers);

        Function sm3 = fnAt(SM3_DRV);
        println("\n=== SM3 driver (consumer) = "
            + (sm3!=null? sm3.getEntryPoint():"?") + " ===");

        // decompile từng ứng viên -> file + đếm hash-hint
        List<long[]> rank = new ArrayList<long[]>();   // {off, hashHint, cSize, feedsUnhex}
        Map<Long,String> names = new HashMap<Long,String>();
        for (long off : cands) {
            Function f = fnAt(off);
            if (f==null){ println("  0x"+Long.toHexString(off)+": NO FUNCTION"); continue; }
            DecompileResults r = dec.decompileFunction(f, 120, monitor);
            String c = (r!=null && r.decompileCompleted())
                ? r.getDecompiledFunction().getC()
                : "// decompile FAILED: " + (r!=null? r.getErrorMessage() : "null result") + "\n";
            String lc = c.toLowerCase();
            long hh = count(lc,"sha")+count(lc,"sm3")+count(lc,"md5")
                    + count(lc,"digest")+count(lc,"hex");
            long feeds = (c.contains(Long.toHexString(currentProgram.getImageBase()
                    .getOffset()+UNHEX)) || c.contains("891f4"))? 1:0;
            File fp = new File(OUTDIR, "fn_"+Long.toHexString(off)+".c");
            FileWriter w = new FileWriter(fp); w.write(c); w.close();
            names.put(off, f.getName());
            rank.add(new long[]{off, hh, c.length(), feeds});
            println("  0x"+Long.toHexString(off)+" -> "+f.getName()
                +"  (C "+c.length()+"B, hash-hint="+hh+", feeds-unhex="+feeds+")");
        }

        // RANK: feeds-unhex trước, rồi hash-hint
        Collections.sort(rank, new Comparator<long[]>(){
            public int compare(long[] a, long[] b){
                if (a[3]!=b[3]) return Long.compare(b[3],a[3]);
                return Long.compare(b[1],a[1]);
            }});
        println("\n=== RANK (feeds-unhex + hash-hint cao = ứng viên producer) ===");
        for (long[] x : rank)
            println("  0x"+Long.toHexString(x[0])+"  feeds-unhex="+x[3]
                +"  hash-hint="+x[1]+"  size="+x[2]+"  "+names.get(x[0]));
        println("\nĐọc _ghidra_out/fn_*.c của hàm top-rank = producer slot16.");
        dec.dispose();
    }

    static long count(String hay, String needle){
        long n=0; int i=0;
        while ((i=hay.indexOf(needle,i))>=0){ n++; i+=needle.length(); }
        return n;
    }
}

File: huongB_devirt19/ghidra_find_slot16.py
# Ghidra headless postScript — khoanh vùng slot16 producer.
# Chạy:
#   $GHIDRA/support/analyzeHeadless <projDir> tt \
#     -import huongB_devirt19/bin/libmetasec_ov.so \
#     -postScript ghidra_find_slot16.py \
#     -scriptPath huongB_devirt19
# Mục tiêu: unhex decoder @0x891f4 có 5 caller; hàm nào build hex slot16 (32 char)
# thì nó chính là producer. Script decompile 5 ứng viên + dump C + xref graph.
#@category TikTok
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

# offset file (= vaddr) -> địa chỉ Ghidra (imageBase + offset)
BASE = currentProgram.getImageBase()
def A(off):
    return BASE.add(off)

UNHEX      = 0x891f4
CANDIDATES = [0x887e0, 0xcaa0c, 0x119108, 0x1349ac, 0x13ab30]
SM3_DRV    = 0x9fdac   # consumer slot16 (SM3 driver) — để nhận diện path đúng
OUTDIR     = os.path.join(os.path.dirname(getSourceFile().getAbsolutePath()), "_ghidra_out")
try: os.makedirs(OUTDIR)
except: pass

fm  = currentProgram.getFunctionManager()
dec = DecompInterface(); dec.openProgram(currentProgram)
mon = ConsoleTaskMonitor()

def fn_at(off):
    f = fm.getFunctionContaining(A(off))
    return f

def decompile(f):
    r = dec.decompileFunction(f, 120, mon)
    if r and r.decompileCompleted():
        return r.getDecompiledFunction().getC()
    return "// decompile FAILED"

def callers_of(off):
    out = []
    for ref in getReferencesTo(A(off)):
        if ref.getReferenceType().isCall():
            fc = fm.getFunctionContaining(ref.getFromAddress())
            out.append((ref.getFromAddress(), fc.getEntryPoint() if fc else None))
    return out

print("=== image base = %s ===" % BASE)
print("=== callers of UNHEX 0x%x ===" % UNHEX)
for site, ent in callers_of(UNHEX):
    print("  call@%s  in fn %s" % (site, ent))

# decompile 5 ứng viên -> file, và tìm dấu hiệu producer:
#   heuristic: hàm producer sẽ (a) gọi 1 hàm hash/SM3 trước khi build hex,
#              (b) tạo std::string rồi feed vào unhex,
#              (c) nằm trên path tới SM3_DRV consumer.
sm3_fn = fn_at(SM3_DRV)
sm3_ep = sm3_fn.getEntryPoint() if sm3_fn else None
print("\n=== SM3 driver (consumer) fn = %s ===" % sm3_ep)

report = []
for off in CANDIDATES:
    f = fn_at(off)
    if not f:
        print("  0x%x: NO FUNCTION" % off); continue
    c = decompile(f)
    fp = os.path.join(OUTDIR, "fn_%x.c" % off)
    with open(fp, "w") as fh: fh.write(c)
    # đếm callee là hash-ish + độ dài
    ncalls = c.count("(")
    hashish = sum(c.lower().count(k) for k in ("sha","sm3","md5","digest","hex"))
    feeds_unhex = ("0x%x" % UNHEX) in c or "FUN_%08x" % (BASE.getOffset()+UNHEX) in c
    report.append((off, len(c), hashish, str(f.getName())))
    print("  0x%x -> %s  (C %dB, hash-hint=%d) dumped %s" %
          (off, f.getName(), len(c), hashish, fp))

print("\n=== RANK (nhiều hash-hint = khả năng producer cao) ===")
for off, sz, hh, nm in sorted(report, key=lambda r: -r[2]):
    print("  0x%x  hash-hint=%d  %s" % (off, hh, nm))
print("\nĐọc _ghidra_out/fn_*.c: tìm hàm build chuỗi 32 hex feed vào unhex = producer.")
dec.dispose()
